### PR TITLE
PD-829 dRAID Primer Article Content Changes

### PR DESCRIPTION
**Pull request by micjohnson777**
This commit adds more information on the dRAID variables, and a place holder for the calculators.

It adds a caveats section and an expand with use cases for dRAID.

Thanks for contributing to TrueNAS documentation! By opening a Pull Request, you're acknowledging that your changes will be distributed under the [Creative Commons 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

### CODE REVIEW

File: content/References/dRAIDprimer.md [hunk 1]
```diff
@@ -12,12 +15,18 @@ It does this by building the dRAID vdev from multiple internal raid groups that
 Depending on data block size and compression requirements, a dRAID pool could have significantly less total storage, especially in situations where large numbers of small data blocks are being stored.
 
 {{< hint type="tip" title="dRAID Usage Recommendations" >}}
-Due to concerns with storage efficiency, dRAID vdev layouts are only recommended in very specific situations where the TrueNAS storage array has numerous (>100) attached disks that are expected to fail frequently and the array is storing large data blocks.
+Due to concerns with storage efficiency, dRAID vdev layouts are only recommended in very specific situations where the TrueNAS storage array has numerous (>100) attached disks that are expected to fail frequently and the array is storing large data blocks. 
+If deploying on SSDs, dRAID can be a viable option for high-performance large-block workloads such as video production and some HPC storage, but test the configuration thoroughly before putting it into production.
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:19]:
```suggestion
If deploying on SSDs, dRAID can be a viable option for high-performance large-block workloads, such as video production and some HPC storage, but test the configuration thoroughly before putting it into production.
```

File: content/References/dRAIDprimer.md [hunk 2]
```diff
@@ -12,12 +15,18 @@ It does this by building the dRAID vdev from multiple internal raid groups that
...
 
-Current investigations between dRAID and RAIDZ vdev layouts find that RAIDZ layouts store data more efficiently in all general use case scenarios, and especially in scenarios where small blocks of data are being stored.
+Current investigations between dRAID and RAIDz vdev layouts find that RAIDZ layouts store data more efficiently in all general use case scenarios, and especially where small blocks of data are being stored.
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:21]:
```suggestion
Current investigations between dRAID and RAIDz vdev layouts find that RAIDZ layouts store data more efficiently in all general use case scenarios and especially where small blocks of data are being stored. 
```

File: content/References/dRAIDprimer.md [hunk 3]
```diff
@@ -12,12 +15,18 @@ It does this by building the dRAID vdev from multiple internal raid groups that
...
+If deploying on SSDs, dRAID can be a viable option for high-performance large-block workloads such as video production and some HPC storage, but test the configuration thoroughly before putting it into production.
 
-Current investigations between dRAID and RAIDZ vdev layouts find that RAIDZ layouts store data more efficiently in all general use case scenarios, and especially in scenarios where small blocks of data are being stored.
+Current investigations between dRAID and RAIDz vdev layouts find that RAIDZ layouts store data more efficiently in all general use case scenarios, and especially where small blocks of data are being stored. 
+dRAID is not suited to applications with primarily small-block data reads and writes, such as VMware and databases that are better suited to mirror and RAIDz vdevs.
+
+In general, consider dRAID where having reduced and greatly-improve resilver time and returning pools to a healthy state faster is needed.
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:24]:
```suggestion
In general, consider dRAID where having greatly-reduced resilver time and returning pools to a healthy state faster is needed.
```

File: content/References/dRAIDprimer.md [hunk 4]
```diff
@@ -12,12 +15,18 @@ It does this by building the dRAID vdev from multiple internal raid groups that
...
+The implementation of dRAID is very new and has not been tested to the same extent as RAIDZ.
+If you storage project cannot tolerate risk consider waiting until dRAID becomes more well-established and widely tested.
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:26]:
```suggestion
If you storage project cannot tolerate risk, consider waiting until dRAID becomes more well-established and widely tested.
```

File: content/References/dRAIDprimer.md [hunk 5]
```diff
@@ -81,48 +90,97 @@ This is essentially a resilver, but data redistributes across the pool to meet t
...
+| Distributed hot spare (*S*) | Unlike in a RAIDz configuration where spares remain inactive until needed, in a dRAID configuration spare capacity is distributed across the drives. This results in all drives being active members of the pool. This number cannot change after pool creation. |
+| Parity Level (*P*) | Distributed parity level of a dRAID redundancy group. This ranges from 1 to 3 and is similar to the RAIDz parity level. |
+| Redundancy group | dRAID layouts use equivalent RAIDz vdevs as the foundation for the complete dRAID vdev. A redundancy group is composed of parity devices and data devices. Redundancy group size impacts storage performance and capacity. |
 | Vdev | An OpenZFS virtual device. dRAID layouts allow much larger vdevs. 100+ device vdevs are not uncommon. Distributed hot spares are shared across all redundancy groups in a vdev. |
-
 {{< /truetable >}}
 
 ## dRAID Capacity Estimations
 
+A single dRAID vdev can have multiple redundancy groups just as a RAIDz can have multiple vdevs.
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:119]:
```suggestion
A single dRAID vdev can have multiple redundancy groups, just as a RAIDz can have multiple vdevs.
```

File: content/References/dRAIDprimer.md [hunk 6]
```diff
@@ -81,48 +90,97 @@ This is essentially a resilver, but data redistributes across the pool to meet t
...
+A dRAID redundancy group is roughly the equivalent of a RAIDz vdev.
+dRAID redundancy groups can be much wider than RAIDz vdevs and still have the same level of redundancy. dRAID variables in estimating layouts are:
+
+* Parity (P) in each redundancy group
+* Data disk (D) in each redundancy group
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:124]:
```suggestion
* Data disks (D) in each redundancy group
```

File: content/References/dRAIDprimer.md (hunk 3 above)
Comment (chrismillsapsix) [inline content/References/dRAIDprimer.md:24]: greatly-improved

File: content/References/dRAIDprimer.md [hunk 7]
```diff
@@ -81,48 +90,97 @@ This is essentially a resilver, but data redistributes across the pool to meet t
...
+<div style="text-align: right; font-size: smaller; padding-top: 1em;">
+    dRAID Visualizer originally created by <a href="https://jro.io/truenas/draid/">Jason Rose</a>.
+</div>
+
+## Caveats
+
+dRAID layouts require more testing before we make layout recommendations on the number of spares or parity levels.
+When using dRAID, consider these factors:
+
+* dRAID does not support partial-strip writes.
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:167]:
```suggestion
* dRAID does not support partial-stripe writes.
```

File: content/References/dRAIDprimer.md [hunk 8]
```diff
@@ -45,24 +54,24 @@ When datasets are expected to have a heavy sequential I/O pattern, a 1MB record
...
 
-#### Distributed hot spare
+#### Distributed Hot Spare
 
-dRAID uses a different methodology from raidz to provide hot spare capacity to the data vdev.
+dRAID uses a different methodology from RAIDz to provide hot spare capacity to the data vdev.
 Distributed hot spares are allocated as blocks on each disk in the pool.
 This means that hot spares must be allocated during vdev creation, cannot be modified later, and that every disk added to the vdev is active within the storage pool.
 
 Disk failure results in the dRAID vdev copying the parity and data blocks onto the spare blocks in the each fixed stripe width disk.
```
Comment (chrismillsapsix) [inline content/References/dRAIDprimer.md:71]: "spare blocks in the each fixed stripe width disk." This part reads poorly, "in the each" needs to be changed.

File: content/References/dRAIDprimer.md [hunk 9]
```diff
@@ -81,48 +90,97 @@ This is essentially a resilver, but data redistributes across the pool to meet t
...
+
+* dRAID performance matches that of an equivalent RAIDz pool.
+  Vdev IOPS scale based on the quantity of redundancy groups per row and throughput scales based on the quantity of data disks per row.
+
+* If creating a dRAID pool with more than 255 disks, you need to use multiple vdevs with fewer disks rather than one vdev with the maximum number of 255.
+  Adhering to best practices means keeping all vdevs alike. So rather than taking 300 disks and creating one 255-wide vdev and one 45-wide vdev, use two 150-wide or three 100-wide vdevs to improve performance and redundancy.
+  
+* Use dRAID redundancy groups with similar numbers of drives and parity levels as in RAIDz vdevs.
+  
+* dRAID pool expansion requires the addition of one or more new vdevs just as with RAIDz.
```
Comment (DjP-iX) [inline content/References/dRAIDprimer.md:182]:
```suggestion
* dRAID pool expansion requires the addition of one or more new vdevs, just as with RAIDz.
```

File: content/References/dRAIDprimer.md (hunk 7 above)
Comment (chrismillsapsix) [inline content/References/dRAIDprimer.md:167]: stripe

**Review by DjP-iX — CHANGES_REQUESTED** [review]

File: content/References/dRAIDprimer.md [hunk 10]
```diff
@@ -81,48 +90,97 @@ This is essentially a resilver, but data redistributes across the pool to meet t
...
+When using dRAID, consider these factors:
+
+* dRAID does not support partial-strip writes.
+
+* Small-block writes to dRAID vdev are padded with zeros to span a redundancy group, which creates a lot of wasted space.
+  Include special allocations class vdevs (i.e., L2ARCs, SLOGs, and metadata) in a dRAID layout to handle this type of IO.
+
+* You cannot add virtual spares after creating the vdev because spare disks are distributed into the dRAID vdev.
+
+* dRAID performance matches that of an equivalent RAIDz pool.
```
Comment (chrismillsapsix) [inline content/References/dRAIDprimer.md:174]: I do not know that we have sufficient empiral data that make any performance claims at this point. dRAID performancce _should_ scale similarly to RAIDz, but until performance can back that up with data I think we should avoid making authoritive statesments related to performance.

File: content/References/dRAIDprimer.md (hunk 8 above)
Comment (micjohnson777) [reply content/References/dRAIDprimer.md:71]: removed "the" which should not be here in this sentence so it now reads "in each fixed stripe width disk"

File: content/References/dRAIDprimer.md (hunk 10 above)
Comment (micjohnson777) [reply content/References/dRAIDprimer.md:174]: Modified this statement.

File: content/References/dRAIDprimer.md (hunk 7 above)
Comment (micjohnson777) [reply content/References/dRAIDprimer.md]: fixed in next commit.

**Review by DjP-iX — APPROVED** [review]

**Pull request merged**